pkg/steps/bundle_source.go

```diff
@@ -69,35 +70,56 @@ func (s *bundleSourceStep) run(ctx context.Context) error {
 	return handleBuild(ctx, s.buildClient, build, s.artifactDir)
 }
 
-func replaceCommand(pullSpec, with string) string {
-	return fmt.Sprintf("find . -type f -name \\*.yaml -exec sed -i 's?%s?%s?g' {} +", pullSpec, with)
+func replaceCommand(pullSpec, with string) []string {
+	sedSub := fmt.Sprintf("s?%s?%s?g", pullSpec, with)
+	return []string{`find`, `.`, `-type`, `f`, `-regex`, `.*\.\(yaml\|yml\)`, `-exec`, `sed`, `-i`, sedSub, `{}`, `+`}
 }
 
 func (s *bundleSourceStep) bundleSourceDockerfile() (string, error) {
 	var dockerCommands []string
-	dockerCommands = append(dockerCommands, "")
 	dockerCommands = append(dockerCommands, fmt.Sprintf("FROM %s:%s", api.PipelineImageStream, api.PipelineImageStreamTagReferenceSource))
-	is, err := s.imageClient.ImageStreams(s.jobSpec.Namespace()).Get(context.TODO(), api.StableImageStream, meta.GetOptions{})
-	if err != nil {
-		return "", fmt.Errorf("failed to get stable imagestream: %w", err)
-	}
+	imagestreams := make(map[string]*imageapi.ImageStream)
 	for _, sub := range s.config.Substitutions {
-		replaceSpec, exists := util.ResolvePullSpec(is, sub.With, false)
+		imageSubSplit := strings.Split(sub.With, ":")
+		if len(imageSubSplit) != 2 {
+			return "", fmt.Errorf("image to be substituted into manifests is invalid (does not contain `:`")
+		}
+		streamName := imageSubSplit[0]
+		tagName := imageSubSplit[1]
+		var is *imageapi.ImageStream
+		if stream, ok := imagestreams[streamName]; ok {
+			is = stream
+		} else {
+			stream, err := s.imageClient.ImageStreams(s.jobSpec.Namespace()).Get(context.TODO(), streamName, meta.GetOptions{})
+			if err != nil {
+				return "", fmt.Errorf("failed to get %s imagestream: %w", streamName, err)
+			}
+			imagestreams[streamName] = stream
+			is = stream
+		}
+		replaceSpec, exists := util.ResolvePullSpec(is, tagName, true)
 		if !exists {
-			return "", fmt.Errorf("failed to get replacement imagestream for image tag `%s`", sub.With)
+			return "", fmt.Errorf("failed to get replacement imagestream and tag for image `%s`", sub.With)
 		}
-		// The \ character has to be escaped in the dockerfile to run correctly
-		replace := strings.ReplaceAll(replaceCommand(sub.PullSpec, replaceSpec), `\`, `\\`)
-		dockerCommands = append(dockerCommands, fmt.Sprintf(`RUN ["bash", "-c", "%s"]`, replace))
+		// format command for dockerfile
+		dockerFormattedCommand := `"` + strings.Join(replaceCommand(sub.PullSpec, replaceSpec), `" "`) + `"`
+		dockerCommands = append(dockerCommands, fmt.Sprintf(`RUN %s`, dockerFormattedCommand))
 	}
-	dockerCommands = append(dockerCommands, "")
 	return strings.Join(dockerCommands, "\n"), nil
 }
 
-func (s *bundleSourceStep) Requires() []api.StepLink {
-	return []api.StepLink{
-		api.InternalImageLink(api.PipelineImageStreamTagReferenceSource),
+func getPipelineSubs(subs []api.PullSpecSubstitution) []api.StepLink {
```

Comment:
Why only pipeline? We have `api.LinkForImage` to be used with `config.DependencyParts` - please see the multi-stage or dependency code